<a href="https://colab.research.google.com/github/amf60/Vac_Covid19_BR_SC_Combinations_of_Concern_4/blob/main/Combinations_of_Concern_sc04082021_10am.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<Font color = 'brown'>**Covid-19 vaccination records in the Brazilian state of Santa Catarina**

<Font color = 'brown'>**Vac_Covid19_BR_SC_Combinations_of_Concern_4**


*   Clinical Data Scientist: Aldir MEDEIROS FILHO
*   FLORIANOPOLIS, [Santa Catarina(SC)](https://en.wikipedia.org/wiki/Santa_Catarina_(state), Brasil
*   04 Aug 2021


**Introduction**:

Per [Alura Bootacamp Applied Data Sciences II plan](https://www.alura.com.br/bootcamp/data-science-aplicada/matriculas-abertas), there was no project to be delivered for the Module 4.

Therefore this project was created as my personal bonus for Module 4 (out 6) of this Alura bootcamp. 

It is a continuation and extension of the exploratory work I already done for the first 3 modules of this bootcamp, when I reviewed the OpenDataSus registers of covid19 vaccination in the Brazilian State of Santa Catarina(SC) up to 02 July 2021. The documentation of work done on the previous 3 modules can be find in the followint github links:

* https://github.com/amf60/Vac_Covid19_BR_SC_Combinations_of_Concern_1
* https://github.com/amf60/Vac_Covid19_BR_SC_Combinations_of_Concern_2
* https://github.com/amf60/Vac_Covid19_BR_SC_Combinations_of_Concern_3

The main findings from this initial work done first using a mix of terminal command line, Python&Colab and Excel for the first two modules and then all the 3 phases in one notebook for module 3 were:

> A trend of ~ 12% of registers at end of May and June, that we can not confirm which vaccine an individual received when I cross-checked 3 variables from the master dataset. I named this troublesome combinations as <Font color = 'brow'>'Combinations of Concern' - CoC </Font> (step 5 here after for definition).

> This trend of ~ 12% proportion of CoC's registrations was already noticed back in March and April, when the nr of registers were less than 1 Million and so I could use only Excel.

<Font color = 'green'>***DES-EXCELANDO!!!!***

<font color='blue'>My objectives for this personal Module 4 project are:



*   <font color='blue'>to keep using only Google Colab notebook and Python related packages, to check the OpenDataSus registration;
*   <font color='blue'>verify if the 12 % trend of 'Combinations of Concern' is still a matter for concern at end of July 2021.

This project is a work in progress and is being conceived and designed with the end-user in mind: the SC DIVE's UDVE managers and local vaccination teams accross the state of Santa Catarina. 

My final goal is that local UDVE teams may use the delivers of the final bootcamp project to help correct, clean and update their covid 19 registration files. 

I am also aiming a high level of automation of the entire monitoring process, so the end-users can, at any time, quickly and easily track progress of the corrective actions to address the 'Combinations of Concern' - CoC.

# <font color = 'blue'>**Phase 1 - importing master dataset and some data wrangling**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib as mpl
#import seaborn as sns

print('Versão do pandas -> %s' % pd.__version__)
print('Versão do numpy -> %s' % np.__version__)
print('Versão do matplotlib -> %s' % mpl.__version__)
#print('Versão do seaborn -> %s' % sns.__version__)
#print('Versão do scipy -> %s' % scipy.__version__)

Versão do pandas -> 1.1.5
Versão do numpy -> 1.19.5
Versão do matplotlib -> 3.2.2


# <Font color = 'green'>**Step 1** - imported the raw master csv dataset into the Google Drive folder created for this project.

This is the fastest process I found to upload file into Colab (less than 2 minutes for a 2.6 GB master file).

However when the file was inside the drive folder, when right clicked on it, I could not find the path to this file using pd.read_csv(''). 

To resolve it I used tips from this [source](https://neptune.ai/blog/google-colab-dealing-with-files). That helped me transfer the file from the drive folder to Colab Files, as shown in the Alura teachings, and then to copy path to add to pd.read_csv.

# <Font Color = 'green'>**Step 2** - Imported the raw master file into Google Colab, renamed the headers and some data cleaning

Renamed all 34 original colum's head & variable names (shorter and identifiable by column nr). For the original colunm names please check this [source](https://opendatasus.saude.gov.br/dataset/b772ee55-07cd-44d8-958f-b12edd004e0b/resource/38ead83d-b115-4219-852e-7244792bc311/download/dicionario-de-dados-vacinacao.pdf).

To have the most updated dataset up to 31 July 2021, I downloaded the file from [OpenDataSus](https://opendatasus.saude.gov.br/dataset/covid-19-vacinacao/resource/ef3bd0b8-b605-474b-9ae5-c97390c197a8?view_id=6bf433e7-16ec-4dfa-a725-08db02322bd6) on 04 Aug 21 at 10 am, and re-named my downloaded master file to match this (sc0408_10am_csv).

Per the original source records at the time of the download, the latest updated of the original database was done on Tuesday, 04 Aug 21 at 00:15 am. I hope, we will have most of the updated registrations up to Friday, 31 Jul 2021 to compare with the data from  end June 2021 used for modules 1, 2 and 3.

used tip from this [link](https://www.geeksforgeeks.org/pandas-strip-whitespace-from-entire-dataframe/) to clear whitespace 

In [ ]:
sc0408_10am = pd.read_csv('/content/drive/MyDrive/Alura DS Bootcamp/Combinations of Concern Projects/Vac_C19_CoC_4_M4_AluraDSB_sc040821_10am/sc040821_10am.csv', sep =';', header = 0, skipinitialspace = True, names=['c1_document_id','c2_p_id','c3_p_idade','c4_p_dn','c5_p_sex','c6_p_r_codigo','c7_p_r_valor','c8_p_ibgemunicipio','c9_p_pais','c10_p_nmmunicipio','c11_p_endereco_nmpais','c12_p_uf','c13_p_cep','c14_p_enumnacionalidade','c15_e_valor','c16_e_razaosocial','c17_e_nofantasia','c18_e_municipio_codigo','c19_e_municipio_nome', 'c20_e_uf', 'c21_v_grupoatendimento_codigo','c22_v_grupoatendimento_nome','c23_v_categoria_codigo','c24_v_categoria_nome','c25_v_lote', 'c26_v_fabricante_nome','c27_v_fabricante_referencia','c28_v_dataaplicacao','c29_v_descricao_dose','c30_v_codigo','c31_v_nome','c32_sistema_origem','c33_data_importacao_rnds','c34_id_sistema_origem'])

In [ ]:
sc0408_10am.shape

(4854329, 34)

#<font color = 'brown'>IMPORTANT REMINDERS BEFORE WE MOVE ON: 

 

1. For information: 1.721.370 news registers were added between 30 Jun 21 (3,133,344 registrations) and 04 Aug 21. 

2. Each row, or register, correspond to the application of one dose, either Dose 1 or Dose 2. Therefore the total of 4.854.329 registrations in this dataset does not mean 4.854.329 vaccinated people in Santa Catarina.

3.   Per oficial [DIVE's boletim published on 03 Aug 21](http://www.dive.sc.gov.br/conteudos/boletim2021/corona53/boletim-epidemiologico-04-08.pdf) the distribution betweeen doses 1 and dose 2 is:

*   dose 1 -      3.163.712 (%)
*   dose 2 -      1.148.033 (%)
*   Single dose -   201.278 ()
*   Total -       4.513.023 (so, not as updated as our dataset)


4. The four covid19 vaccines used in Santa Catarina (SC) until 30 Jun 21 are:

| Vaccine <br>Manufacturers 	| Nr of <br>doses 	| Actual <br>doses interval 	| Interval changed <br>during<br> the vaccination plan <br>from 	| Vaccine available<br> in SC since? 	|
|:-:	|:-:	|:-:	|:-:	|:-:	|
| Coronavac_Butantan 	| 2 	| 21 to 28 days 	| 14 Days 	| 18 Jan 21 	|
| AztraZeneca 	| 2 	| 120 days 	| 28 days 	| 23 Jan 21 	|
| Pfizer 	| 2 	| 120 days 	| No 	| 02 May 21 	|
| Janssen 	| 1 	| NA 	| NA 	| 24 Jun 21 	|


The above table was easily created with a tool from this [source](https://www.tablesgenerator.com/markdown_tables) as suggested by tip from this [one](https://towardsdatascience.com/cheat-sheet-for-google-colab-63853778c093)

For further information about the change of the vacination intervals [Folha de Sao Paulo - 05 May 2021](https://www1.folha.uol.com.br/equilibrioesaude/2021/05/veja-quais-sao-os-intervalos-entre-as-doses-das-vacinas-contra-covid-no-brasil-e-por-que-eles-sao-diferentes.shtml)

In [ ]:
sc0408_10am.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4854329 entries, 0 to 4854328
Data columns (total 34 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   c1_document_id                 object 
 1   c2_p_id                        object 
 2   c3_p_idade                     float64
 3   c4_p_dn                        object 
 4   c5_p_sex                       object 
 5   c6_p_r_codigo                  float64
 6   c7_p_r_valor                   object 
 7   c8_p_ibgemunicipio             float64
 8   c9_p_pais                      float64
 9   c10_p_nmmunicipio              object 
 10  c11_p_endereco_nmpais          object 
 11  c12_p_uf                       object 
 12  c13_p_cep                      float64
 13  c14_p_enumnacionalidade        object 
 14  c15_e_valor                    int64  
 15  c16_e_razaosocial              object 
 16  c17_e_nofantasia               object 
 17  c18_e_municipio_codigo         int64  
 18  c1

In [ ]:
sc0408_10am.head()

c1_document_id  ... c34_id_sistema_origem
0  2d11e103-8172-419a-bfba-d1bb09d4524e-i0b0  ...                 16341
1  b3ea27d9-5bf6-42cb-98e9-28e948062930-i0b0  ...                 16341
2  cab1dd0a-378a-4ee5-ab2b-af7334dadfc5-i0b0  ...                 16341
3  bb127a9c-8819-4ee1-bdc7-0e5b4005cc10-i0b0  ...                 16341
4  090583ba-e8bd-47ba-9901-a9907d483c51-i0b0  ...                 16341

[5 rows x 34 columns]

## <Font color = 'green'>*Converted the 3 date's columns from object to datetime.*

Tips for the conversion of c33 taken from this [link](https://www.statology.org/convert-datetime-to-date-pandas/)

In [ ]:
sc0408_10am['c4_p_dn'] = pd.to_datetime(sc0408_10am['c4_p_dn'])
sc0408_10am['c28_v_dataaplicacao'] = pd.to_datetime(sc0408_10am['c28_v_dataaplicacao'])
sc0408_10am['c33_data_importacao_rnds'] = pd.to_datetime(sc0408_10am['c33_data_importacao_rnds']).dt.date
sc0408_10am['c33_data_importacao_rnds'] = pd.to_datetime(sc0408_10am['c33_data_importacao_rnds']).dt.normalize()

In [ ]:
sc0408_10am.tail(3)

c1_document_id  ... c34_id_sistema_origem
4854326  c0ad2add-50aa-449f-8360-fce187efe4ef-i0b0  ...                 18602
4854327  7c844019-b8e6-4f19-a406-2857f223a8ca-i0b0  ...                 16341
4854328  b58ac131-05df-42c0-9e24-9c9fe329216f-i0b0  ...                 16341

[3 rows x 34 columns]

In [ ]:
sc0408_10am.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4854329 entries, 0 to 4854328
Data columns (total 34 columns):
 #   Column                         Dtype         
---  ------                         -----         
 0   c1_document_id                 object        
 1   c2_p_id                        object        
 2   c3_p_idade                     float64       
 3   c4_p_dn                        datetime64[ns]
 4   c5_p_sex                       object        
 5   c6_p_r_codigo                  float64       
 6   c7_p_r_valor                   object        
 7   c8_p_ibgemunicipio             float64       
 8   c9_p_pais                      float64       
 9   c10_p_nmmunicipio              object        
 10  c11_p_endereco_nmpais          object        
 11  c12_p_uf                       object        
 12  c13_p_cep                      float64       
 13  c14_p_enumnacionalidade        object        
 14  c15_e_valor                    int64         
 15  c16_e_razaosoci

##<font color = 'green'>*Eliminated empty space and shortned the code for variable column 'c29_v_descricao_dose'*

In [ ]:
sc0408_10am.loc[sc0408_10am['c29_v_descricao_dose'] == '    1ª Dose', 'c29_v_descricao_dose'] = 'D1'
sc0408_10am.loc[sc0408_10am['c29_v_descricao_dose'] == '    2ª Dose', 'c29_v_descricao_dose'] = 'D2'
sc0408_10am.head()

c1_document_id  ... c34_id_sistema_origem
0  2d11e103-8172-419a-bfba-d1bb09d4524e-i0b0  ...                 16341
1  b3ea27d9-5bf6-42cb-98e9-28e948062930-i0b0  ...                 16341
2  cab1dd0a-378a-4ee5-ab2b-af7334dadfc5-i0b0  ...                 16341
3  bb127a9c-8819-4ee1-bdc7-0e5b4005cc10-i0b0  ...                 16341
4  090583ba-e8bd-47ba-9901-a9907d483c51-i0b0  ...                 16341

[5 rows x 34 columns]

## <Font color = 'green'>*Sorting the entire dataset based on dates of the variable 'c28_v_dataaplicacao'*

In [ ]:
sc0408_10am.sort_values(by='c28_v_dataaplicacao', ascending=True, inplace=True)
sc0408_10am

c1_document_id  ... c34_id_sistema_origem
837682   c0622be3-b4a1-4e4f-9c21-8e0e0cc7f7da-i0b0  ...                 18602
3435720  4af998fc-ba82-40e6-bfd4-d2fcde965c29-i0b0  ...                 18602
2810301  d041ac6e-fdfd-4875-944e-6d2c4cc79191-i0b0  ...                 18602
2655507  dffd702e-cd88-4e7c-8795-e4e0a7784af2-i0b0  ...                 18562
3452470  c9c8dc1f-d435-423d-94d0-5f68fe9b00e9-i0b0  ...                 18602
...                                            ...  ...                   ...
3859532  e301087e-84d3-4a39-a724-027af2c89c73-i0b0  ...                 16341
262345   d92aec42-469a-4d90-8b86-be7fc9fd118e-i0b0  ...                 16341
3859534  603eda2e-4335-40b7-afc0-a4efbccd5fcc-i0b0  ...                 16341
3859526  30ce7135-b42a-4218-a120-33696935e023-i0b0  ...                 16341
262351   3b1baad5-d8e2-4ec3-bddf-c9f0696ec7c9-i0b0  ...                 16341

[4854329 rows x 34 columns]

# <Font color ='blue'> **Phase 2 - Adding a new column in preparation for the dataset for analysis**

# <Font color = 'green'>**Step 3** _ Adding a column identifying the UDVE of each register (row) in the dataset.

SC has 295 counties. [DIVE](http://www.dive.sc.gov.br), the local state instituition in charge of the distribution and tracking of the covid19 vaccines applications, shared the 295 in 17 UDVEs (Decentralized Epidemiological Surveillance Units in english).

Usually DIVE refers to its UDVEs by the name of the main city in the region an UDVE covers. 

To facilitate the coding for this project I renamed the UDVEs with a number, where 01 starts on the most western side of the state of SC and double digits ones the eastern (or costal) regions. 

These numbers tried to match the number used for regionales by another state institution helping coordinate the covid19 vaccinations in the state, called [COSENS SC](https://www.cosemssc.org.br/regioes_de_saude/). 

Tried because as you may notice, DIVE is organised in 17 UDVE regions and COSEMS SC in 16 regions and their regions borders not always match. How that helps, I do not know.

<Font color = 'green'>Based on tip from this [source](https://stackoverflow.com/questions/63558830/how-to-create-a-new-column-in-pandas-and-add-value-to-that-new-column-based-on-t) managed to succesfully create a new column labelled 'c35_SC_UDVE_Nr', where each register got an identification for its UDVE based on the county code where the vaccine shot was applied.

In [ ]:
def get_udve(row):
    if row ['c18_e_municipio_codigo'] in [420080,420208,420209,420215,420490,420500,420535,420640,420660,420765,420840,421100,421200,421223,421415,421507,421520,421555,421625,421670,421720,421875]:
        return 'udve01'
    elif row['c18_e_municipio_codigo'] in [420010,420253,420445,420517,420530,420560,420768,420917,420945,421055,421185,421227,421340,421610,421910,421950,421970]:
        return 'udve02'
    elif row['c18_e_municipio_codigo'] in [420050,420055,420165,420257,420310,420350,420410,420420,420435,420440,420470,420475,420543,420665,420775,420785,420895,421050,421090,421140,421145,421165,421187,421210,421290,421315,421420,421535,421568,421569,421575,421600,421690,421715,421730,421755,421775,421795,421885]:
        return 'udve03'
    elif row['c18_e_municipio_codigo'] in [420075,420127,420430,420760,420770,420780,420800,420985,421260,421310,421390,421750,421960]:
        return 'udve04'
    elif row['c18_e_municipio_codigo'] in [420005,420040,420287,420360,420390,420400,420415,420520,420670,420680,420860,420900,420920,421003,421105,421180,421790,421850,421915,421917,421985]:
        return 'udve05'
    elif row['c18_e_municipio_codigo'] in [420160,420300,420315,420480,420550,420555,420675,420757,420970,421005,421070,421300,421335,421440,421540,421550,421605,421825,421930]:
        return 'udve06'
    elif row['c18_e_municipio_codigo'] in [420213,420330,420380,420790,420810,421010,421030,421110,421220,421360,421500,421580,421830]:
        return 'udve07'
    elif row['c18_e_municipio_codigo'] in [420450,420650,420890,421060,421740]:
        return 'udve08'
    elif row['c18_e_municipio_codigo'] in [420130,420205,420210,420580,420845,420910,421620,421635]:
        return 'udve09'
    elif row['c18_e_municipio_codigo'] in [420200,420245,420320,420710,420820,420830,421000,421130,421250,421280,421350]:
        return 'udve10'
    elif row['c18_e_municipio_codigo'] in [420020,420030,420180,420190,420285,420419,420510,420690,420740,420850,420915,420950,420990,421085,421270,421370,421400,421410,421450,421460,421480,421530,421567,421780,421860,421920,421935,421940]:
        return 'udve11'
    elif row['c18_e_municipio_codigo'] in [420125,420170,420220,420240,420270,420290,420515,420590,420630,420750,421320,421470,421510,421820]:
        return 'udve12'
    elif row['c18_e_municipio_codigo'] in [420060,420070,420090,420110,420120,420230,420370,420540,420570,420600,420980,421020,421150,421190,421230,421430,421570,421590,421630,421660,421725,421800]:
        return 'udve13'
    elif row['c18_e_municipio_codigo'] in [420100,420243,420250,420260,420325,420340,420417,420455,420930,421175,421189,421205,421330,421505,421650,421680,421890,421895]:
        return 'udve14'
    elif row['c18_e_municipio_codigo'] in [420150,420280,420395,420610,420620,420720,420730,420880,420940,421240,421265,421490,421545,421560,421700,421710,421840,421870]:
        return 'udve15'
    elif row['c18_e_municipio_codigo'] in [420425,420460,420545,420700,420960,421120,421160,421170,421760,421835,421900,42200]:
        return 'udve16'
    elif row['c18_e_municipio_codigo'] in [420140,420195,420207,420519,420870,421040,421080,421125,421225,421380,421565,421640,421770,421810,421880]:
        return 'udve17'


In [ ]:
sc0408_10am['c35_SC_UDVE_Nr'] = sc0408_10am.apply(get_udve, axis = 1)

In [ ]:
sc0408_10am_udve = sc0408_10am

In [ ]:
sc0408_10am_udve.head()

c1_document_id  ... c35_SC_UDVE_Nr
837682   c0622be3-b4a1-4e4f-9c21-8e0e0cc7f7da-i0b0  ...         udve13
3435720  4af998fc-ba82-40e6-bfd4-d2fcde965c29-i0b0  ...         udve13
2810301  d041ac6e-fdfd-4875-944e-6d2c4cc79191-i0b0  ...         udve07
2655507  dffd702e-cd88-4e7c-8795-e4e0a7784af2-i0b0  ...         udve07
3452470  c9c8dc1f-d435-423d-94d0-5f68fe9b00e9-i0b0  ...         udve13

[5 rows x 35 columns]

In [ ]:
sc0408_10am_udve.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4854329 entries, 837682 to 262351
Data columns (total 35 columns):
 #   Column                         Dtype         
---  ------                         -----         
 0   c1_document_id                 object        
 1   c2_p_id                        object        
 2   c3_p_idade                     float64       
 3   c4_p_dn                        datetime64[ns]
 4   c5_p_sex                       object        
 5   c6_p_r_codigo                  float64       
 6   c7_p_r_valor                   object        
 7   c8_p_ibgemunicipio             float64       
 8   c9_p_pais                      float64       
 9   c10_p_nmmunicipio              object        
 10  c11_p_endereco_nmpais          object        
 11  c12_p_uf                       object        
 12  c13_p_cep                      float64       
 13  c14_p_enumnacionalidade        object        
 14  c15_e_valor                    int64         
 15  c16_e_razao

#<font color = 'blue'>**Phase 3 - Identifying and formally defining 'Combinations of Concern' (CoC)**

# <Font Color='green'>**Step 4** - Identifying and counting the different combinations between the 4 variables that allow identication of which covid19 vaccine an individual received.

| Column name                               | Column name in English    | comment                       |
|-------------|-----------------------------|---------------------------|
|<Font color ='brown'>c26_v_fabricante_nome | c26_vaccine manufacturer  | 
|<Font color ='brown'>c30_v_codigo          | c30_v_code                |
|<Font color ='brown'>c31_v_nome            | c31_v_name                |
|<Font color ='brown'>c32_sistema_origem    | c32_source_system         | system used at local level to enter registration data


The process used in this step was done based on the learning gathered from this [source](https://stackoverflow.com/questions/35268817/unique-combinations-of-values-in-selected-columns-in-pandas-data-frame-and-count)

In [ ]:
v_fabri_codigo_nome_sysorig_sc0408_10am_udve = sc0408_10am_udve[['c26_v_fabricante_nome','c30_v_codigo','c31_v_nome', 'c32_sistema_origem',]].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4854329 entries, 837682 to 262351
Data columns (total 4 columns):
 #   Column                 Dtype 
---  ------                 ----- 
 0   c26_v_fabricante_nome  object
 1   c30_v_codigo           int64 
 2   c31_v_nome             object
 3   c32_sistema_origem     object
dtypes: int64(1), object(3)
memory usage: 185.2+ MB


In [ ]:
v_fabri_codigo_nome_sysorig_sc0408_10am_udve = sc0408_10am_udve[['c26_v_fabricante_nome','c30_v_codigo','c31_v_nome','c32_sistema_origem']]
v_fabri_codigo_nome_sysorig_sc0408_10am_udve

c26_v_fabricante_nome  ...             c32_sistema_origem
837682   SERUM INSTITUTE OF INDIA LTD  ...  ESUS APS - NACIONAL (OFFLINE)
3435720  SERUM INSTITUTE OF INDIA LTD  ...  ESUS APS - NACIONAL (OFFLINE)
2810301  SERUM INSTITUTE OF INDIA LTD  ...  ESUS APS - NACIONAL (OFFLINE)
2655507               SINOVAC/BUTANTA  ...                      IDS Saúde
3452470  SERUM INSTITUTE OF INDIA LTD  ...  ESUS APS - NACIONAL (OFFLINE)
...                               ...  ...                            ...
3859532         FUNDACAO OSWALDO CRUZ  ...                       Novo PNI
262345            MINISTERIO DA SAUDE  ...                       Novo PNI
3859534           MINISTERIO DA SAUDE  ...                       Novo PNI
3859526         FUNDACAO OSWALDO CRUZ  ...                       Novo PNI
262351              FUNDACAO BUTANTAN  ...                       Novo PNI

[4854329 rows x 4 columns]

In [ ]:
v_fabri_codigo_nome_sysorig_sc0408_10am_udve.groupby(['c26_v_fabricante_nome','c30_v_codigo','c31_v_nome','c32_sistema_origem']).size()

c26_v_fabricante_nome                    c30_v_codigo  c31_v_nome                                                 c32_sistema_origem               
ASTRAZENECA AB                           85            Vacina Covid-19 - Covishield                               G-MUS                                   2494
                                                                                                                  G-MUS - Gestão Municipal de Saúde      10414
ASTRAZENECA FARMACEUTICA DO BRASIL LTDA  85            Vacina Covid-19 - Covishield                               IPM istemas LTDA                        3813
BUTANTAN                                 86            Covid-19-Coronavac-Sinovac/Butantan                        IPM istemas LTDA                         341
BioNTech/Fosun Pharma/Pfizer             87            Vacina covid-19 - BNT162b2 - BioNTech/Fosun Pharma/Pfizer  ESUS APS - NACIONAL (OFFLINE)          75977
FUNDACAO BUTANTAN                        86            Co

# <Font color = 'green'>**Step 5** - Defining and creating a new column ['c_36_Combination_of_Concern'] that identify and classify 'combinations of concern' based on their frequency.

#<Font color = 'brown'>**Definition of Combination of Concern(CoC)'**: within an individual vaccination registration, any combination between the 4 variables above mentioned that would raise doubt about which covid vaccine an individual received. 

Per example: 'SERUM INSTITUTE OF INDIA LTD' is ONE OF the manufacturers across the globe for AztraZeneca(AZ) covid 19 related vaccines. AZ covid19 vaccines produced by SERUM are known as ['COVISHIELD'](https://www.bbc.com/news/health-57695835).  That's inconsistent with a 'c31_v_nome' = 'Covid-19-Coronavac-Sinovac/Butantan' in the same register, as this last one is manufactured by Coronavac China and/or Butantan Brasil.

For more details about identification of covid19 vaccine manufacturers and vaccine names please check references at these [ANVISA source]((https://www.gov.br/anvisa/pt-br/assuntos/paf/coronavirus/vacinas)) re-accessed on 26 Jun 2021.

# <font color ='brown'> Coding 'Combinations of Concern' based on their frequency.

*   <Font color = 'brown'>**CoC_1** = Very High Frequency - more than 100,000 registers
*   <Font color = 'brown'>**CoC_2** = High Frequency - between 10,000 and 100,000 registers
*   <Font color = 'brown'>**CoC_3** = 'Low' Frequency - less than 10,000 registers
*   <Font color = 'brown'>**'no_issue'** = At best knowledge of this author no remarkble issue identified at the time this project was prepared ( 30 June 2021)

<font color = 'brown'>**Author's note:** Strictly speaking, from the Pharmacovigilance point of view, 'no issue' is not strictly correct. But it will fulfill for the moment the purpose of this part of the work. More details in the conclusion at the end.

The set of codes here after I was inspired by tips from this [source](https://towardsdatascience.com/the-ultimate-guide-for-column-creation-with-pandas-dataframes-83b8c565110e)

In [ ]:
def combination_of_concern(row):
    if row['c26_v_fabricante_nome'] == 'SERUM INSTITUTE OF INDIA LTD' and row['c30_v_codigo'] == 86 and row['c31_v_nome'] == 'Covid-19-Coronavac-Sinovac/Butantan':
        return 'CoC_1'
    elif row['c26_v_fabricante_nome'] == 'SINOVAC LIFE SCIENCE CO LTD' and row['c30_v_codigo'] == 85 and row['c31_v_nome'] == 'Vacina Covid-19 - Covishield':
        return 'CoC_1'
    elif row['c26_v_fabricante_nome'] == 'MINISTERIO DA SAUDE' and row['c30_v_codigo'] == 89 and row['c31_v_nome'] == 'Covid-19-AstraZeneca ':
        return 'CoC_1'
    elif row['c26_v_fabricante_nome'] == 'SINOVAC/BUTANTA' and row['c30_v_codigo'] == 85 and row['c31_v_nome'] == 'Vacina Covid-19 - Covishield':
        return 'CoC_3'
    elif row['c26_v_fabricante_nome'] == 'FUNDACAO OSWALDO CRUZ' and row['c30_v_codigo'] == 86 and row['c31_v_nome'] == 'Covid-19-Coronavac-Sinovac/Butantan':
        return 'CoC_3'
    elif row['c26_v_fabricante_nome'] == 'FUNDACAO OSWALDO CRUZ' and row['c30_v_codigo'] == 88 and row['c31_v_nome'] == 'Vacina covid-19 - Ad26.COV2.S - Janssen-Cilag':
        return 'CoC_3'
    elif row['c26_v_fabricante_nome'] == 'OXFORD-ASTRAZENECA' and row['c30_v_codigo'] == 86 and row['c31_v_nome'] == 'Covid-19-Coronavac-Sinovac/Butantan':
        return 'CoC_3'
    elif row['c26_v_fabricante_nome'] == 'OXFORD-ASTRAZENECA' and row['c30_v_codigo'] == 87 and row['c31_v_nome'] == 'Vacina covid-19 - BNT162b2 - BioNTech/Fosun Ph...':
        return 'CoC_3'
    elif row['c26_v_fabricante_nome'] == 'PFIZER/BIONTECH' and row['c30_v_codigo'] == 86 and row['c31_v_nome'] == 'Covid-19-Coronavac-Sinovac/Butantan':
        return 'CoC_3'
    elif row['c26_v_fabricante_nome'] == 'INTERVAX LTDA' and row['c30_v_codigo'] == 85 and row['c31_v_nome'] == 'Vacina Covid-19 - Covishield':
        return 'CoC_3'
    else:
        return 'no_issue'
     

#<font color = 'brown'> When compared to last previous update on 30 June 2021, we notice the following changes in the 'Combination of Concern' categories:

*   <font color = 'blue'> A significant increase of CoC_1 frequency. 
*   <font color = 'blue'> No CoC_2 group in July (there was 2 in May);
*   <font color = 'blue'> A small reduction on the nr of CoC_3 category.

The data above mays suggest that clean up of the database is not happening in systematic and large scale for the CoC_1 category.







In [ ]:
sc0408_10am_udve['c_36_Combination_of_Concern'] = sc0408_10am_udve.apply(combination_of_concern, axis=1)

In [ ]:
sc0408_10am_udve_CoC = sc0408_10am_udve

In [ ]:
sc0408_10am_udve_CoC.c_36_Combination_of_Concern.unique()

array(['CoC_1', 'no_issue', 'CoC_3'], dtype=object)

In [ ]:
sc0408_10am_udve_CoC.tail()

c1_document_id  ... c_36_Combination_of_Concern
3859532  e301087e-84d3-4a39-a724-027af2c89c73-i0b0  ...                    no_issue
262345   d92aec42-469a-4d90-8b86-be7fc9fd118e-i0b0  ...                    no_issue
3859534  603eda2e-4335-40b7-afc0-a4efbccd5fcc-i0b0  ...                    no_issue
3859526  30ce7135-b42a-4218-a120-33696935e023-i0b0  ...                    no_issue
262351   3b1baad5-d8e2-4ec3-bddf-c9f0696ec7c9-i0b0  ...                    no_issue

[5 rows x 36 columns]

# <Font color ='blue'>**Phase 4 - Crosstabing UDVE and CoC at county level**

# <Font color = 'green'>**Step 6** - Crosstabing UDVE and CoC

Tips from Alura curso 'Estatistica com Python parte 1: Frequencias e Medidas - aula 2 - Distribuicao de frequencia + for formating right - [source](https://stackoverflow.com/questions/17232013/how-to-set-the-pandas-dataframe-data-left-right-alignment) and [source](https://pbpython.com/styling-pandas.html)

In [ ]:
udve_CoC = pd.crosstab(sc0408_10am_udve_CoC.c35_SC_UDVE_Nr , sc0408_10am_udve_CoC.c_36_Combination_of_Concern)
def highlight(s):
    if s.CoC_1 > 100000:
        return ['background-color: red'] * 3
    elif (s.CoC_1 > 10000) and (s.CoC_1 < 100000):
        return ['background-color: orange'] * 3
    elif (s.CoC_1 > 1000) and (s.CoC_1 < 10000):
        return ['background-color: yellow'] * 3
    else:
        return ['background-color: white'] * 3

udve_CoC.style.set_properties(**{'text-align': 'right'}).format('{0:,.0f}').apply(highlight, axis=1)


Tip from [source](https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html) (Finer Control: Display Values)

# <font color = 'green'>Proportions of CoCs categories per UDVE 

tip [source](https://stackoverflow.com/questions/50820659/compute-row-percentages-in-pandas-dataframe)

UDVEs 13, 7, 6, 9, 3 and 10 are, respectively, the ones of greatest concern due to the potential large cluster of unknown vaccination status (CoC1), based on the data of this dataset.

<Font color = 'red'>Percentual - a great example of how 'data cleaning' in clinical data sciences can be a very sensitive matter. CoC_3 is low in frequency but potentially high in consequencies ('No one is safe, until everyone is'). Low percentage does not mean 'low risk'.

In [ ]:
Per_udve_CoC = udve_CoC[['CoC_1','CoC_3','no_issue']] = udve_CoC[['CoC_1','CoC_3','no_issue']].apply(lambda x: x/x.sum(), axis=1) * 1

Per_udve_CoC.style.set_properties(**{'text-align': 'right'}).format("{:.2%}")

# <Font color = 'green'>**Step 07** - Crosstabing UDVE and CoC, per county, for the two UDVES with the highest frequency of 'Combinations of Concern'(CoC): UDVEs 13 and 07 respectively.

# <font color ='red'>**UDVE 13**

In [ ]:
udve_13 = sc0408_10am_udve_CoC.loc[sc0408_10am_udve_CoC['c35_SC_UDVE_Nr'] == 'udve13']
udve_13.shape


(774494, 36)

In [ ]:
udve_13_CoC = pd.crosstab(udve_13.c19_e_municipio_nome , udve_13.c_36_Combination_of_Concern)
udve_13_CoC.style.set_properties(**{'text-align': 'right'}).format('{0:,.0f}')

#<Font color = 'orange'>**UDVE 07**

In [ ]:
udve_07 = sc0408_10am_udve_CoC.loc[sc0408_10am_udve_CoC['c35_SC_UDVE_Nr'] == 'udve07']
udve_07.shape

(227175, 36)

In [ ]:
udve_07_CoC = pd.crosstab(udve_07.c19_e_municipio_nome , udve_07.c_36_Combination_of_Concern)
udve_07_CoC.style.set_properties(**{'text-align': 'right'}).format('{0:,.0f}')

# <Font Color = 'blue'> **PHASE 5** - 'Time Serie' Exploration

#<Font color='green'> **Step 08** - Counting total nr of registers per month

Tips from [source](https://stackoverflow.com/questions/38792122/how-to-group-and-count-rows-by-month-and-year-using-pandas)

In [ ]:
c_28 = sc0408_10am_udve_CoC['c28_v_dataaplicacao'].groupby([sc0408_10am_udve_CoC['c28_v_dataaplicacao'].dt.year.rename('year'), sc0408_10am_udve_CoC['c28_v_dataaplicacao'].dt.month.rename('month')]).agg({'count'})
c_28

count
year month         
2020 12           3
2021 1        70317
     2       172782
     3       613956
     4       809429
     5       729720
     6      1019784
     7      1427763
     8        10575

# <Font Color = 'green'>**Step 9** - Checking 'Combination of Concern' frequency since start of covid19 vaccination in Santa Catarina as per dataset sc0408_10am.

Tips from [source](https://stackoverflow.com/questions/44908383/how-can-i-group-by-month-from-a-date-field-using-python-pandas)

In [ ]:
c28_36 = sc0408_10am_udve_CoC.groupby([(sc0408_10am_udve_CoC['c28_v_dataaplicacao'].dt.strftime('%B')),'c_36_Combination_of_Concern']).size()
c28_36

c28_v_dataaplicacao  c_36_Combination_of_Concern
April                CoC_1                           121957
                     CoC_3                                8
                     no_issue                        687464
August               no_issue                         10575
December             CoC_1                                3
February             CoC_1                            34525
                     CoC_3                               12
                     no_issue                        138245
January              CoC_1                            12607
                     CoC_3                                9
                     no_issue                         57701
July                 CoC_1                           105555
                     no_issue                       1322208
June                 CoC_1                            92074
                     no_issue                        927710
March                CoC_1                         

<Font color = 'red'> Here I could try to find out how to pivot the table above to the wished result here after (done manually). However, I enjoy this 'sensorial tête_à_tête' with the data as it provides very interesting insights. 

Per example we notice that: 


*   Wrong registrations entered for December 2020 are still not corrected (vaccination in SC only started on 18 Jan 21);
*   When we compare data from end June with end July we notice that frequency for category CoC_1 is still raising for the early months of January, February and March, but reducing for April, May and June.

In [ ]:
sc0408_10am_c_28_c_36 = {
    'Month':    ['December','January','February', 'March', 'April','May','June','July'],
    'CoC_1':     [3, 12607, 34525, 127466, 121957, 81883, 92074,105555],
    'CoC_2':     [0, 0, 0, 0, 0, 0, 0, 0],
    'CoC_3':     [0, 9, 12, 3, 8, 1, 0, 0],
    'no_issue':  [0, 57701, 138245, 486487, 687464, 647836, 927710, 1322208]
    }

In [ ]:
sc0408_10am_df_c_28_c_36 = pd.DataFrame(data=sc0408_10am_c_28_c_36)
sc0408_10am_df_c_28_c_36

Month   CoC_1  CoC_2  CoC_3  no_issue
0  December       3      0      0         0
1   January   12607      0      9     57701
2  February   34525      0     12    138245
3     March  127466      0      3    486487
4     April  121957      0      8    687464
5       May   81883      0      1    647836
6      June   92074      0      0    927710
7      July  105555      0      0   1322208

#<font color = 'green'>**Step 10** - Recalling data from 30 Jun 2021 for comparisons with data from up to 30 June 2021:

In [ ]:
sc0207_07am_c_28_c_36 = {'Month':['December', 'January',  'February', 'March', 'April',     'May',    'Jun'],
                    'CoC_1':   [3,           11834,       34385,   112652,   103964,   58107,    47241],
                    'CoC_2':   [0,               0,           0,        0,        0,       0,        0],
                    'CoC_3':   [0,               9,          13,      437,        6,       2,        0],
                 'no_issue':   [0,           56505,      137148,   484289,   680737,  629570,   776442]}

In [ ]:
sc0207_07am_c_28_c_36 = pd.DataFrame(data=sc0207_07am_c_28_c_36)
sc0207_07am_c_28_c_36

Month   CoC_1  CoC_2  CoC_3  no_issue
0  December       3      0      0         0
1   January   11834      0      9     56505
2  February   34385      0     13    137148
3     March  112652      0    437    484289
4     April  103964      0      6    680737
5       May   58107      0      2    629570
6       Jun   47241      0      0    776442

# **<Font color='blue'>Brief Comments and conclusions:**

*   <font color = 'green'>**DES_EXCELEI !!!** </font>- With the help of the Alura Bootcamp DS 2, Aluras courses Programming with Python, Python for DataSciences, services as stackoverflow, medium and others, I accomplished my objective to deliver this module project entirely within one Google Colab notebook using Python related packages.

*   We notice that, in Santa Catarina, **on 31 JUL 21, a small reduction (from 12% to 10%(467103 (CoC_1 + CoC_3) out of a total of 4854329 registrations)) on the previous the overall trend of 12% registrations of covid_19 vaccinations with 'Combinations of Concern'.**

*   UDVE 13 alone clusters 82% of the state total of 'Combinations of Concern' (398908 out of 467103). It seems a significant increase as end of June this % was 62%

*   "'*No_issue*'" - as already mentioned, from the pharmacovigilance point of view, is not totally correct, as I noticed some other 'combinations of worrying' (to distinguish of the definition of 'CoC'). So I may tackle 'combinations of worrying' in the projects of the next two modules of this bootcamp.

*   Finally, as explained, I am tailoring all the above with the end_user in mind: the SC UDVE manager and local teams at the each 'vaccination center'. Therefore, at the end of this bootcamp we may use the material above to easily identify and isolate all the registrations with 'combinations of concern' for appreciation at several levels - UDVE, county, vaccination center, etc. More about how that can be done in a sensitive way with efficiente Risk Based Monitoring techniques I use in clinical trials may be presented in the next two modules.











